### PR TITLE
Include files in vfs on demand

### LINKED ISSUE #10178: Spurious failure to load a file from `OUT_DIR`
I have a project (https://gitlab.com/subplot/subplot) which has a `build.rs` which writes a file out to be included into the build.

The file is written out to `OUT_DIR` (`embedded_files.inc`)

In my code I have:

```rust
static EMBEDDED_FILES: &[(&str, &[u8])] = include!(concat!(env!("OUT_DIR"), "/embedded_files.inc"));
```

In VSCode, I have that line have the following error shown:

```
failed to load file `/home/dsilvers/dev-git/subplot/target/debug/build/subplot-a90f1b6ad6ba44a6/out/embedded_files.inc`
rust-analyzer(macro-error)
```

However, that file very much does exist, so the error is spurious.  Is there a way to get rust-analyzer to find this file and stop complaining?

### LINKED ISSUE #11025: Go to definition / open link for file path in `include_str` should work for non-Rust files
Split off from #11019. See https://github.com/rust-analyzer/rust-analyzer/issues/11019#issuecomment-994966731.

Go to definition / open link for file paths in `include`, `include_str`, `include_bytes` _does_ work already (https://github.com/rust-analyzer/rust-analyzer/issues/11019#issuecomment-994479218).

Sadly, this only works if the file extension is `rs` (even if the actual file content is valid Rust code). My codebase on the other hand `include_str`s JS and CSS files and `include_bytes` WOFF files (just like rustdoc). For them, this feature does not work.

I would say more often than not one doesn't `include_str` Rust files (contrary to `include`). I am not sure if that still counts as _jump to definition_ in the  LSP sense but it's reasonable to support as _open link_ in my opinion. I am not familiar with LSP though.

**rust-analyzer version**: rust-analyzer version: db2a7087b 2021-12-13 stable

**rustc version**: rustc 1.59.0-nightly (48a5999fc 2021-12-01)

### LINKED ISSUE #11205: No completions in pattern position of `for` loop with leading `&`
```rs
for &$0 in () {}
```
We do not get any completions in this position whatsoever, most likely due to parser recovery not working here

### LINKED ISSUE #12181: Bug: Inlay hint type annotation for includes_bytes! macro displays size 0 rather than the correct size
The type hint for a variable that is the result of the include_bytes! macro shows size zero rather than its correct size (which is known at compile time)

Screen Shot 2022-05-06 at 2 19 02 PM

However after typing the exact hint in as it is presented, RA shows an error indicating a type mismatch on the size of the array.

image

I would expect the hint to display `: &[u8; 122880]`.

**rust-analyzer version**: (eg. output of "Rust Analyzer: Show RA Version" command)
rust-analyzer version: 5dce1ff02 2022-05-02 stable

**rustc version**: (eg. output of `rustc -V`)
rustc 1.60.0 (7737e0b5c 2022-04-04)

**relevant settings**: (eg. client settings, or environment variables like `CARGO`, `RUSTUP_HOME` or `CARGO_HOME`)

### PR DESCRIPTION
**Pull request by HKalbasi**
fix #12181
fix #11025
fix #10178

I'm not sure if this is in line with `vfs` abstraction or not, but I think we can start with it.

### CODE REVIEW

File: crates/vfs-notify/src/lib.rs [hunk 1]
```diff
@@ -24,10 +24,14 @@ pub struct NotifyHandle {
     _thread: jod_thread::JoinHandle,
 }
 
+impl UnwindSafe for NotifyHandle {}
```
Comment (HKalbasi) [inline crates/vfs-notify/src/lib.rs:27]: Is it ok?

**Comment by HKalbasi** [comment]
I think CI failure is unrelated to this PR. It also happens on #13275

**Comment by Veykril** [comment]
Ye CI fails because the proc-macro server doesnt support 1.64 which released yesterday

**Comment by Veykril** [comment]
Sorry that I didn't get back to this. I was pondering about this change a bit(in part because I am still not too familiar with our VFS). I think there is a problem here now though, whenever a path won't be resolvable from a given source root (so not because the file ending isn't tracked but because it might lie outside the source root, think `..` segments), this will now subscribe that path over and over again as the source root will keep on never resolving the path.

**Comment by HKalbasi** [comment]
So the problem is that we may push a file multiple times in `watched_entries` and it would be solved if we check existence before pushing, right? Would it be only specific to paths outside the source root or would it happen with non existent files inside of source root as well?

**Comment by Veykril** [comment]
Not quite, what I mean is something like the following scenario, we have a module like `foo/src/lib.rs` which has an include for `include!("../../../some_other/path.rs")`. This include's path may lie outside of the sourceroot containing `foo`. Now with the logic here, we'll try to resolve the path from the include, fail and therefor subscribe it. This will trigger a recalculation of the `FileSet`s and `SourceRoot`s, and afterwards as this new path lies outside the sourceroot of `foo`, it will end up in another sourceroot so we will try to resubscribe that path again, as the `foo` sourceroot still cant resolve it.

This is basically https://github.com/rust-lang/rust-analyzer/issues/9173. Now the issue here specifically is about fixing non rs-files but the way this is implemented it will try to register files from different sourceroots and fail. So either we try to fix the sourceroot problem as well (which is gonna be very tough because we do not want to break the concept of source roots so this would require a lot of though), or we try to limit subscribing to only unknown files that are still within our current sourceroot, though I don't know how simple that would be either (although I assume that to be simpler).

**Comment by HKalbasi** [comment]
I found a more fundamental problem. This PR in current state actually doesn't work, since `watched_entries` and `watcher` fields in `NotifyActor` don't do anything. They are either dead code or are for a specific config or usages without LSP. Normally, watching is on the client, and LSP doesn't have a request (or I didn't find) for adding a file to watch list. This makes me wonder, how bad it would be to include all files (with some exceptions, for example target directory) without explicit demand?

**Comment by Veykril** [comment]
Right .. the reason for that I think is that we load the data of all files in the VFS into memory, so if we add all files, we'll have a lot of memory wasted on unnecessary files. So I think this kind of lazy loading would probably be the best choice, though as you said i dont think we can even make that work with client watching ...

**Comment by HKalbasi** [comment]
In the last changes, the client will watch all the files, but we will read from disk and store in memory only if the file is in the `watched_entries`. I also fixed the source root problem and now we won't subscribe if the target path belongs to another `FileSet`.

**Comment by HKalbasi** [comment]
@Veykril is [this problem](https://github.com/rust-lang/rust-analyzer/pull/13279#issuecomment-1280503675) now resolved?

**Comment by Veykril** [comment]
Sorry again for having pushed this to side for so long. I am still not too comfortable with this part of the code :) I have to say, I am not too fond of us asking the client to watch every file now, this will include everything in the `target` folder, and rustc creates a ton of files (my rather freshly cleaned r-a checkout already has 17k files in `target` again). That's a lot of unnecessary watches ... and afaik file watching is not necessarily cheap on some systems (though I might be wrong here, happy to be proven wrong).

**Comment by bjorn3** [comment]
Linux has a very low system wide limit on the amount of inodes you can watch.

**Comment by Veykril** [comment]
Ye figured, then this isn't an option I would say

**Comment by HKalbasi** [comment]
I think vscode already watches every file for its own proposes, and it is not unhappy with the `target` directory (there was a non rust project where vscode warned me about too much file to watch, and I had to manually add something (IIRC `node_modules`) to `files.watcherExclude`). And I tested this branch on the rust-analyzer and didn't notice any difference in resource usage, on linux and vscode. It uses 4801 inotify watches with and without this changes (my target directory has 38k items, and I see the target directory in r-a logs if I print `Invalidate` messages, I don't know how it is possible) and cpu/ram usage is normal. So I think it doesn't change anything for vscode, and watching entire directory of a rust project, even big ones like rustc and r-a, is feasible (because vscode can) but I don't know if there are clients with poor implementation of watching.

**Comment by Veykril** [comment]
I am going to close this, I don't think this is the right way to approach this and I'd rather we revisit the VFS/SourceRoot abstraction to see if we can come up with a scheme that works out better here.

**Pull request closed without merging**